### PR TITLE
Update admin data model UI for users with limited permissions

### PR DESCRIPTION
**Pull request by alxnddr**
# Feature-level data model permission

## How to verify
**This is a FE piece only so testing various states requires changing code**

#### As an Admin
- Go to Admin -> Data model
- Ensure it behaves the same way as before, and admins can edit the data model
#### As a regular user
- Ensure you cannot go to the Admin Settings as a regular user

#### As an Admin
- Go to Admin -> Data model
- Ensure it behaves the same way as before, and admins can edit the data model

#### As a regular user without a Data model permission
- Ensure you cannot go to the Admin settings as a regular user

#### As a regular user with Data model permission enabled
**Setup**
1) `enterprise/frontend/src/metabase-enterprise/advanced_permissions/utils.ts` make `canEditEntityDataModel` return `true`
2) `frontend/src/metabase/redux/user.js` grant `can_access_data_model` to the user by changing the reducer:

```
[REFRESH_CURRENT_USER]: {
      next: (state, { payload }) => {
        return {
          ...payload,
          can_access_data_model: true,
        };
      },
    },
```

- Ensure you can go to the Admin Settings and it opens the Data model
- Ensure you can only view databases/tables/schemas data model with `read` data model permission
- Ensure you can edit databases/tables/schemas data model with `write` data model permission

<video src="https://user-images.githubusercontent.com/14301985/154602282-99589123-d2da-410d-8a38-9f6195cc6afc.mov" />

### CODE REVIEW

File: frontend/src/metabase/admin/datamodel/components/database/MetadataTableList/MetadataTableList.tsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  const [search, setSearch] = useState<{ text: string; regex: RegExp | null }>({
+    text: "",
+    regex: null,
+  });
+
+  const handleSearchChange: ChangeEventHandler<HTMLInputElement> = event => {
+    setSearch({
+      text: event.target.value,
+      regex: event.target.value
+        ? new RegExp(regexpEscape(event.target.value), "i")
```
Comment (ranquild) [inline frontend/src/metabase/admin/datamodel/components/database/MetadataTableList/MetadataTableList.tsx:44]: It seems that `regex` is not state, and should be computed and memoized from `text`

File: frontend/src/metabase/admin/datamodel/components/database/MetadataTablePicker/MetadataTablePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+export default _.compose(
+  Tables.loadList({
+    query: (_state: State, { databaseId }: MetadataTablePickerProps) => ({
+      dbId: databaseId,
+      include_hidden: true,
+    }),
+    reloadInterval: (
+      _state: State,
+      _props: MetadataTablePickerProps,
+      tables = [],
```
Comment (ranquild) [inline frontend/src/metabase/admin/datamodel/components/database/MetadataTablePicker/MetadataTablePicker.tsx:73]: There should be a type for `Table` too in our api types

File: frontend/src/metabase/admin/datamodel/containers/MetadataEditorApp.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-nocheck
```
Comment (ranquild) [inline frontend/src/metabase/admin/datamodel/containers/MetadataEditorApp.tsx:2]: Can it be removed and `any` used instead where needed?

File: enterprise/frontend/src/metabase-enterprise/advanced_permissions/utils.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+import { Database } from "metabase-types/types/Database";
+import { Table } from "metabase-types/types/Table";
+
+export const canAccessSettings = (user: User) => canAccessDataModel(user);
+
+export const canAccessDataModel = ({ can_access_data_model }: User) =>
+  can_access_data_model;
+
+export const canEditEntityDataModel = (entity: Table | Database | null) =>
+  entity?.data_model_permission == null ||
```
Comment (ranquild) [inline enterprise/frontend/src/metabase-enterprise/advanced_permissions/utils.ts:11]: Why does this check for `null` exist? I couldn't figure it out

File: frontend/src/metabase-types/api/user.ts [hunk 5]
```diff
@@ -12,4 +12,5 @@ export interface User {
   last_login: string;
   has_invited_second_user: boolean;
   personal_collection_id: number;
+  can_access_data_model?: boolean;
```
Comment (ranquild) [inline frontend/src/metabase-types/api/user.ts:15]: Is it really optional? I think that the backend will always return this field. Could you make it required and set to false in the mock creator function?

File: frontend/src/metabase-types/types/Database.ts [hunk 6]
```diff
@@ -36,7 +35,8 @@ export type Database = {
   features: DatabaseFeature[];
   is_full_sync: boolean;
   is_sample: boolean;
-  native_permissions: DatabaseNativePermission;
+  native_permissions: DataAccessPermission;
+  data_model_permission?: DataAccessPermission;
```
Comment (ranquild) [inline frontend/src/metabase-types/types/Database.ts:39]: The same question - is it truly optional, or will the backend always set, e.g. to `none`? If it's optional, how is it different to `none` or `write`?

**Review by ranquild — CHANGES_REQUESTED** [review]
It's still possible to navigate to table settings via keyboard, and some of the settings are writable 
Screenshot 2022-03-25 at 16 48 18
Screenshot 2022-03-25 at 16 48 46

Also it seems wrong that it's possible to discard cached values when you are effectively a read-only user
Screenshot 2022-03-25 at 16 47 39
Screenshot 2022-03-25 at 16 47 50

All comments above are nits

**Comment by alxnddr** [comment]
Closed in favor of https://github.com/metabase/metabase/pull/20952

**Pull request closed without merging**